src/procedures/procedure.c

```diff
@@ -132,12 +130,8 @@ bool Procedure_ContainsOutput(const ProcedureCtx *proc, const char *output) {
 
 bool Proc_ReadOnly(const char *proc_name) {
 	assert(__procedures);
-	ProcGenerator gen = raxFind(__procedures, (unsigned char *)proc_name,
-	  			strlen(proc_name));
-	if(gen == raxNotFound) return false; // Invalid procedure specified, handled elsewhere.
-	/* TODO It would be preferable to be able to determine whether a procedure is read-only
-	 * without creating its entire context; this is wasteful. */
-	ProcedureCtx *ctx = gen(NULL, NULL);
+	ProcedureCtx *ctx = Proc_Get(proc_name);
+	if(!ctx) return false;
```

Comment:
Should this become an assertion?
ASSERT(ctx != NULL) ?

Reply:
no because you might not get a context

Reply:
@DvirDukhan  Who's calling this function?

Reply:
@swilly22  AST_ReadOnly
```c
bool AST_ReadOnly(const cypher_astnode_t *root) {
	// Check for empty query
	if(root == NULL) return true;
	cypher_astnode_type_t type = cypher_astnode_type(root);
	if(type == CYPHER_AST_CREATE                      ||
	   type == CYPHER_AST_MERGE                  ||
	   type == CYPHER_AST_DELETE                 ||
	   type == CYPHER_AST_SET                    ||
	   type == CYPHER_AST_CREATE_NODE_PROPS_INDEX ||
	   type == CYPHER_AST_DROP_NODE_PROPS_INDEX) {
		return false;
	}
	// In case of procedure call which modifies the graph/indices.
	if(type == CYPHER_AST_CALL) {
		const char *proc_name = cypher_ast_proc_name_get_value(cypher_ast_call_get_proc_name(root));
		return Proc_ReadOnly(proc_name);
	}
	uint num_children = cypher_astnode_nchildren(root);
	for(uint i = 0; i < num_children; i ++) {
		const cypher_astnode_t *child = cypher_astnode_get_child(root, i);
		if(!AST_ReadOnly(child)) return false;
	}
	return true;
}
```

Reply:
So we're dealing with a situation where the user specified a none existing procedure and when this test is performed `Proc_ReadOnly` will return false i.e. the procedure is writing, which will flag the query as a write query. it would have been better if we could have set an error at this point, or at least treat the missing procedure as read only, as we want to avoid write contention.

Reply:
@swilly22 
fixed